controllers/tokenupload_controller.go

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/logs"
+
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-logr/logr"
+
+	"github.com/redhat-appstudio/service-provider-integration-operator/pkg/spi-shared/tokenstorage"
+	"k8s.io/apimachinery/pkg/types"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	spi "github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const tokenSecretLabel = "spi.appstudio.redhat.com/upload-secret"
+const spiTokenNameLabel = "spi.appstudio.redhat.com/token-name"
+const providerUrlLabel = "spi.appstudio.redhat.com/providerUrl"
+
+// TokenUploadReconciler reconciles a Secret object
+type TokenUploadReconciler struct {
+	client.Client
+	Scheme       *runtime.Scheme
+	TokenStorage tokenstorage.NotifyingTokenStorage
+}
+
+func (r *TokenUploadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	lg := log.FromContext(ctx)
+
+	list := &corev1.SecretList{}
+
+	// list secrets labeled with "spi.appstudio.redhat.com/upload-secret"
+	err := r.List(ctx, list, client.HasLabels{tokenSecretLabel})
+	if err != nil {
+		lg.Error(err, "can not get list of secrets")
+		return ctrl.Result{}, fmt.Errorf("can not get list of secrets: %w,", err)
+	}
+
+	for i, s := range list.Items {
+
+		var accessToken spi.SPIAccessToken
+
+		// we immediatelly delete the Secret
+		err := r.Delete(ctx, &list.Items[i])
+		if err != nil {
+			logError(ctx, s, fmt.Errorf("can not delete the Secret: %w", err), r, lg)
+			continue
+		}
+
+		// if spiTokenName field is not empty - try to find SPIAccessToken by it
+		if s.Labels[spiTokenNameLabel] != "" {
```

Comment:
would be nice to have logic split to functions from the beginning

Reply:
please propose what exactly you have in mind